nomos-da/subnets/src/proto/mod.rs

```diff
@@ -0,0 +1,64 @@
+use std::io;
+
+use bytes::Bytes;
+use futures::AsyncReadExt;
+use prost::Message;
+
+pub mod dispersal;
+
+const MAX_MSG_LEN_BYTES: usize = 2;
+
+pub fn pack_message(message: &impl Message) -> Result<Vec<u8>, io::Error> {
+    let data_len = message.encoded_len();
+
+    if data_len > (1 << (MAX_MSG_LEN_BYTES * 8)) {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "Message too large",
+        ));
+    }
+
+    let mut buf = Vec::with_capacity(MAX_MSG_LEN_BYTES + data_len);
+    buf.extend_from_slice(&(data_len as u16).to_be_bytes());
+    message.encode(&mut buf).unwrap();
+
+    Ok(buf)
+}
+
+pub async fn unpack_from_reader<M, R>(reader: &mut R) -> Result<M, io::Error>
+where
+    M: Message + Default,
+    R: AsyncReadExt + Unpin,
+{
+    let mut length_prefix = [0u8; MAX_MSG_LEN_BYTES];
+    reader.read_exact(&mut length_prefix).await?;
+    let data_length = u16::from_be_bytes(length_prefix) as usize;
+
+    let mut data = vec![0u8; data_length];
+    reader.read_exact(&mut data).await?;
+    M::decode(Bytes::from(data)).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+}
```

Comment:
Nitpick: I have the feeling that this can be hook to serde somehow. Or is this standard for prost?

Reply:
Protobufs are just a way to serialize messages, nothing else. It's up to the user to build a protocol out of it.
Prost already provides ser/de for message types via `encode`/`decode` methods. The packing that is done here is just to know how many bytes to pull from the reader when data is coming from a wire. As prost is only responsible for ser/de, we need to take care of the "what to parse" and "when" ourselves.

Reply:
> Protobufs are just a way to serialize messages, nothing else. It's up to the user to build a protocol out of it. Prost already provides ser/de for message types via `encode`/`decode` methods. The packing that is done here is just to know how many bytes to pull from the reader when data is coming from a wire. As prost is only responsible for ser/de, we need to take care of the "what to parse" and "when" ourselves.

Uhm, I would like to understand this properly as I may be missing something.But, I do not really think this is the best approach, but we can work on it together later on.